File: clevertap-android-sdk/src/main/java/com/clevertap/android/sdk/CTInAppBasePartialFragment.java
package com.clevertap.android.sdk;

import android.app.FragmentManager;
import android.app.FragmentTransaction;


public abstract class CTInAppBasePartialFragment extends CTInAppBaseFragment {

    @Override
    public void onStart() {
        super.onStart();
        if (isCleanedUp.get()) {
            cleanup();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    void generateListener() {
        if (config != null) {
            setListener(CleverTapAPI.instanceWithConfig(getActivity().getBaseContext(),config));
        }
    }

    @Override
    void cleanup() {
        if (!Utils.isActivityDead(getActivity()) && !isCleanedUp.get()) {
            final FragmentManager fragmentManager = parent.getFragmentManager();
            FragmentTransaction transaction = fragmentManager.beginTransaction();
            try {
                transaction.remove(this).commit();
            } catch (IllegalStateException e) {
                fragmentManager.beginTransaction().remove(this).commitAllowingStateLoss();
            }
        }
        isCleanedUp.set(true);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        didDismiss(null);
    }
}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/fragment/CTInAppBasePartialFragment.kt
package com.clevertap.android.sdk.inapp.fragment

import com.clevertap.android.sdk.CleverTapAPI
import com.clevertap.android.sdk.Utils
import com.clevertap.android.sdk.inapp.InAppDisplayListener
import java.util.concurrent.atomic.AtomicBoolean

internal abstract class CTInAppBasePartialFragment : CTInAppBaseFragment(), InAppDisplayListener {

    private val isCleanedUp = AtomicBoolean()

    override fun onStart() {
        super.onStart()
        if (isCleanedUp.get()) {
            cleanup()
        }
        registerInAppDisplayListener()
    }

    override fun onPause() {
        super.onPause()
        unregisterInAppDisplayListener()
    }

    override fun cleanup() {
        val activity = getActivity()
        if (activity != null && !Utils.isActivityDead(activity)
            && isCleanedUp.compareAndSet(false, true)
        ) {
            val fragmentManager = activity.supportFragmentManager
            val transaction = fragmentManager.beginTransaction()
            try {
                transaction.remove(this).commit()
            } catch (_: IllegalStateException) {
                fragmentManager.beginTransaction().remove(this).commitAllowingStateLoss()
            }
        }
    }


    override fun generateListener() {
        setListener(
            CleverTapAPI.instanceWithConfig(requireContext(), config).coreState.inAppController
        )
    }

    private fun registerInAppDisplayListener() {
        CleverTapAPI.instanceWithConfig(requireContext(), config).coreState.inAppController.registerInAppDisplayListener(this)
    }

    private fun unregisterInAppDisplayListener() {
        CleverTapAPI.instanceWithConfig(requireContext(), config).coreState.inAppController.unregisterInAppDisplayListener()
    }

    override fun hideInApp() {
        didDismiss(null)
    }
}
